### ISSUE TITLE
No support for optional chaining syntax

### ISSUE THREAD

**Issue by niba** [DevExpress/testcafe-hammerhead#2714]
### What is your Scenario?

I have a site which loads a code with the following line

`this._pageLabels?.[this._currentPageNumber - 1] ?? null;`

HH transforms it to the

`__get$(this._pageLabels,this._currentPageNumber-1)  ?? null;`

and this code throws an error during runtime

`Uncaught (in promise) Error: Cannot read property '0' of null
    at Function.t._error (hammerhead.js:12)
    at value (hammerhead.js:12)`

The logic in transformed code is different.
 
I managed to get it working by extending createPropertyGetWrapper / createComputedPropertyGetWrapper with optional flag and changing the logic of method but I don't know if this is a correct solution.

The site is an internal one so I cannot share the url

### What is the Current behavior?

HH transformed code with optional chaining syntax have the different logic than the original one

### What is the Expected behavior?

HH should detect optional chaining and perform a proper transformation

### What is your public website URL? (or attach your complete example)

Private

### What is your TestCafe test code?

I've been using proxy

### Steps to Reproduce

1. Find a site with optional chaining code
2. Try to use proxy on the site
3. Site code breaks

### TestCafe version

testcafe-hammerhead 24.2.1

### Command-line arguments

-

**Comment by felis2803** [comment]
Hello @niba,

Thank you for your report. I have reproduced this bug. Stay tuned for updates in this issue

**Comment by felis2803** [comment]
_**For the team.**_

Example to reproduce:
```html
<button>Click me!</button>
<script>
    function getLastBoxText () {
        const lastBox = this.boxes?.[this.length -1] ?? { text: 'Click me, please!' };

        return lastBox.text;
    }

    const container = {
        boxes:  null,
        length: 0,

        getLastBoxText
    }

    document.querySelector('button').innerText = container.getLastBoxText();
</script>
```

**Comment by raphaelniederer** [comment]
We have the same problem.

Error message in the console:

`````
Error: Cannot read property 'href' of undefined
    at Function.t._error (hammerhead.js:12:31820)
    at t._propertyGetter (hammerhead.js:13:539)
`````

Code:

`export const isLinkEvent = (event?: AnyEventObject): event is LinkEvent => !!event?.data?.href;`

`event` is `undefined` in this case.

We have this issue since Testcafe 10.0.2 - so we are still on 10.0.1

Is there a workaround for this problem?

**Comment by miherlosev** [comment]
Hi @raphaelniederer 

At present, there is no workaround. We will update this thread as soon we have any news.

**Comment by danieltroger** [comment]
Also suffering from this issue.

```javascript
const $42170c0be024d7ea$export$a4f97b27a0592e6c = (url)=>{debugger;return (url ? new URL(url, window.location.origin) : window.location).pathname.split("/").filter((a)=>a).pop()?.replace(/\\.html?$/i, "") || ""}
```

is being transformed to

```javascript
const $42170c0be024d7ea$export$a4f97b27a0592e6c = (url)=>{debugger;return  __call$((url?new URL(url,__get$(window,"location").origin):__get$(window,"location")).pathname.split("/").filter(a=>a).pop(),"replace",[/\.html?$/i,""])  || ""}
```


Which fails with:

```
Error: Cannot call method 'replace' of undefined
    at Function.t._error (hammerhead.js:formatted:15811:23)
    at value (hammerhead.js:formatted:15829:43)
    at $7727f36b03cf11cc$export$8d42ce4827f43582.$42170c0be024d7ea$export$a4f97b27a0592e6c (eval at <anonymous> (hammerhead.js:formatted:15175:41), <anonymous>:1713:76)
    at $7727f36b03cf11cc$export$8d42ce4827f43582.eval (eval at <anonymous> (hammerhead.js:formatted:15175:41), <anonymous>:734:38)
    at $7727f36b03cf11cc$export$eee4c4d924348b05 (eval at <anonymous> (hammerhead.js:formatted:15175:41), <anonymous>:1942:33)
    at new $7727f36b03cf11cc$export$8d42ce4827f43582 (eval at <anonymous> (hammerhead.js:formatted:15175:41), <anonymous>:1932:48)
    at eval (eval at <anonymous> (hammerhead.js:formatted:15175:41), <anonymous>:1966:51)
    at eval (eval at <anonymous> (hammerhead.js:formatted:15175:41), <anonymous>:5018:3)
```

**Comment by danieltroger** [comment]
Is there any documentation on this transformation, why is it necessary? Can I turn it off somehow?

**Comment by felis2803** [comment]
Hello @danieltroger,

This transformation is necessary for hiding some DOM nodes from client scripts of the tested website. TestCafe uses this feature to render its own UI on the tested web page (for example, [this](https://testcafe.io/documentation/402835/guides/basic-guides/debug#client-side-debugging)). It cannot be disabled.

**Comment by danieltroger** [comment]
I don't think the sites I'm testing on would care about extra DOM nodes but thanks for explaining. Opted for switching testcafe to a transpiled version of my script as a workaround.

**Comment by miherlosev** [comment]
> Opted for switching testcafe to a transpiled version of my script as a workaround.

We will not implement new functionality to work around the bug. We will fix the bug.

**Comment by miherlosev** [comment]
Hi @LironHazan 

The cause of the https://github.com/DevExpress/testcafe/issues/6863 is a problem with processing optional chaining syntax.

**Comment by LironHazan** [comment]
@miherlosev which happens when targeting es2020 and be resolved as part of this issue? (just to be sure that I won't need to open another issue, this thread seemed relevant as the transformation is done by this lib)

**Comment by Assafz1983** [comment]
Hi!
Do you happen to have an ETA for this fix?
It blocks us from upgrading our project to es2020.

**Comment by Assafz1983** [comment]
Thanks. I see that you increased the frequency to critical.

Just worth mentioning that it blocks TestCafe users to upgrade their app to use es2020, and hence - it blocks them from onboarding module federation..

**Comment by alex-zh** [comment]
Hello, any updates on this problem? We cannot workaround this problem? 
it's super critical for us because we have tests that test some integration with 3d party system which uses optional chainig, 
so we get errors but cannot rewrite "optional chaninng" code.
The only option - migrate from TestCafe to something else(

**Comment by VasilyStrelyaev** [comment]
We realize that this issue is critical. It is currently marked as such and will be considered of high priority during our planning. Unfortunately, there are no specific estimations I could share at the moment.

**Comment by f5io** [comment]
hi everyone, this became an issue for us after updating sveltekit, none of our tests would run as sveltekit couldn't initialise.

**we've found an interim solution until this is resolved upstream....**

by applying the following patch, with [patch-package](https://www.npmjs.com/package/patch-package), we were able to get our full test suite running :tada:

```diff
diff --git a/node_modules/testcafe-hammerhead/lib/processing/script/transformers/computed-property-get.js b/node_modules/testcafe-hammerhead/lib/processing/script/transformers/computed-property-get.js
index cde9fe1..b461f37 100644
--- a/node_modules/testcafe-hammerhead/lib/processing/script/transformers/computed-property-get.js
+++ b/node_modules/testcafe-hammerhead/lib/processing/script/transformers/computed-property-get.js
@@ -39,6 +39,15 @@ const transformer = {
         // for(object[prop] in source)
         if (parent.type === esotope_hammerhead_1.Syntax.ForInStatement && parent.left === node)
             return false;
+        
+        if (node.property.type === 'PrivateIdentifier')
+            return false
+
+        if (parent.type === 'LogicalExpression' && parent.operator === '??')
+            return false;
+
+        if (parent.type === 'ChainExpression')
+            return false;
         return true;
     },
     run: node => node_builder_1.createComputedPropertyGetWrapper(node.property, node.object)
```

to achieve this for yourself, follow these steps:

1. install `patch-package` as a dev dependency
2. go into `node_modules/testcafe-hammerhead/lib/processing/script/transformers/computed-property-get.js`, and add the following lines to the `condition` function...
```js
if (node.property.type === 'PrivateIdentifier')
   return false

if (parent.type === 'LogicalExpression' && parent.operator === '??')
  return false;

if (parent.type === 'ChainExpression')
  return false;
```
3. run `patch-package testcafe-hammerhead`, this will produce a file `patches/testcafe-hammerhead+24.5.14.patch` which you should commit
4. add a `postinstall` script to your package.json that calls `patch-package`

i would happily open a PR to resolve this issue, but it seems there are types spread out across multiple packages that dont correlate, we were unable to get this to pass typescript type-checking without enforcing `ts-ignore` on the lines added.

**Comment by Aleksey28** [comment]
Hi @f5io,

Thank you for your cooperation. We will research this issue and update this thread once we have any news.

**Comment by LironHazan** [comment]
@f5io Nice!! 
we've **workaround** the issue by a temporary eslint rule which yields errors for` node.optional && node.computed` and transforms it to use lodash instead 
`fixer.replaceText(node, `_.get(${object}, ${property})`)`
But instrumenting a patch is a good idea..

**Comment by f5io** [comment]
I believe there is still an outstanding issue with nullish coalescing that has not been fixed by #2755.

the following js should not blow up during a testcafe run...

```js
const obj = {};
const key = 'foo';
const b = obj[key] ?? 'bar';
```

you could consider member access to 'foo' here optional however the AST does not reflect that in the `MemberExpression` node, as can be seen in this [AST Explorer Link](https://astexplorer.net/#/gist/a8cbd98cc212bfb0c07a916738831b0c/7402c88edacec305987dbe3fcefdadec2e7a610f)

if you would prefer I can open a new issue...

**Comment by Aleksey28** [comment]
Hi @f5io

You are right. This fix doesn't address the `??` operator. Please create a separate issue for the `??` operator. The current issue is related to the optional chaining operator.

**Comment by danieltroger** [comment]
I just glanced at the tests of your MR and didn't see a test for `({ fn: null })?.fn?.();`

Just wanted to ask if you thought about that case?

**Comment by f5io** [comment]
Hi @Aleksey28,

Apologies, on closer inspection it looks like the changes to `_propertyGetter` should have resolved the issues around nullish coalescing also. I could not create a failing test-case against the package. Hopefully this should unblock testcafe against sveltekit applications. Good work all 👍

**Comment by Aleksey28** [comment]
@danieltroger, Thank you for the example. I tried to implement it, and it didn't throw an error. It should work. Please use [this PR](https://github.com/DevExpress/testcafe/pull/6970) to check it.

@f5io, I checked it, and it works without any errors. Thank you for looking into this.

**Comment by AlexanderZhig** [comment]
Hi guys!
I take TC version 1.18.6, that uses testcafe-hammerhead": "24.5.18" and the problem still present: 
I get the following code 
**d=(t&&t["function.from"])??null**
converted to 
**d=t&&t["function.from"]??null**
that leads to 
Uncaught SyntaxError: Unexpected token '??'

**Comment by AlexanderZhig** [comment]
> Hi guys! I take TC version 1.18.6, that uses testcafe-hammerhead": "24.5.18" and the problem is still present: I get the following code **d=(t&&t["function.from"])??null** converted to **d=t&&t["function.from"]??null** that leads to Uncaught SyntaxError: Unexpected token '??'

Any comments on this?

**Comment by miherlosev** [comment]
Hi @AlexanderZhig,

See the future updates related to this issue at [https://github.com/DevExpress/testcafe/issues/6995](https://github.com/DevExpress/testcafe/issues/6995).
